lib/pbench/server/api/resources/users_api.py

```diff
@@ -0,0 +1,534 @@
+            abort(400, message=f"Invalid email: {email_id}")
+        except Exception:
+            self.logger.exception("Exception while registering a user")
+            abort(500, message="INTERNAL ERROR")
+
+
+class Login(Resource):
+    """
+    Pbench API for User Login or generating an auth token
+    """
+
+    def __init__(self, config, logger, auth):
+        self.server_config = config
+        self.logger = logger
+        self.auth = auth
+        try:
+            self.token_expire_duration = self.server_config.get(
+                "pbench-server", "token_expiration_duration"
+            )
+        except NoOptionError:
+            self.token_expire_duration = None
+
+    @Auth.token_auth.login_required(optional=True, f=Auth().verify_auth())
+    def post(self):
+        """
+        Post request for logging in user.
+        The user is allowed to re-login multiple times and each time a new valid auth token will be provided
+
+        This requires a JSON data with required user metadata fields
+        {
+            "username": "username",
+            "password": "password",
+        }
+
+        Required headers include
+
+            Content-Type:   application/json
+            Accept:         application/json
+
+        :return: JSON Payload
+        if we succeed to decrypt the password hash, the returned response_object will include the auth_token
+            response_object = {
+                    "status": "success", # will not present if failed
+                    "message": "Successfully logged in."/"failure message",
+                    "auth_token": "<authorization_token>", # Will not present if failed
+                }
+        """
+        # get the post data
+        post_data = request.get_json()
+        if not post_data:
+            self.logger.warning("Invalid json object: {}", request.url)
+            abort(400, message="Invalid json object in request")
+
+        username = post_data.get("username")
+        if not username:
+            self.logger.warning("Username not provided during the login process")
+            abort(400, message="Please provide a valid username")
+
+        password = post_data.get("password")
+        if not password:
+            self.logger.warning("Password not provided during the login process")
+            abort(400, message="Please provide a valid password")
+
+        try:
+            # fetch the user data
+            user = User.query(username=username)
+        except Exception:
+            self.logger.exception("Exception occurred during user login")
+            abort(500, message="INTERNAL ERROR")
+
+        if not user:
+            self.logger.warning(
+                "No user found in the db for Username: {} while login", username
+            )
+            abort(403, message="No such user, please register first")
+
+        # Validate the password
+        if not check_password_hash(user.password, password):
```

Comment:
what do you mean?

Reply:
Nick means, can a third party with access to the network packets (e.g., by looking with another computer hooked up to the network segment) a  read the password from the request payload?

I presume the answer depends on what protocol we're using.  If we're using HTTPS, then the answer should be "no"; if we're using plain HTTP, then the answer will be "yes".  I presume that the choice of protocol is up to the deployment configuration...is that true?

Reply:
Well, yes, we do; but only because we're still using `http`. We know we can't do that; we just haven't taken on the whole morass of dealing with SSL certificates and signing protocols yet.

We can't do this as a "product" ... we'll need to generate a self-signed cert, and support CA signed certs, and use `https` for everything, at which time nothing will be "in the clear".